src/adapter/src/coord/sequencer/inner/create_materialized_view.rs

```diff
@@ -540,8 +539,26 @@ impl Coordinator {
     ) -> Result<StageResult<Box<CreateMaterializedViewStage>>, AdapterError> {
         // Timestamp selection
         let id_bundle = dataflow_import_id_bundle(global_lir_plan.df_desc(), cluster_id);
-        let (read_holds, dataflow_as_of, storage_as_of, until) =
-            self.select_timestamps(id_bundle, refresh_schedule.as_ref())?;
+
+        let read_holds_owned;
+        let read_holds = if let Some(txn_reads) = self.txn_read_holds.get(session.conn_id()) {
+            // In some cases, for example when REFRESH is used, the preparatory
+            // stages will already have acquired ReadHolds, we can re-use those.
+            tracing::info!(session = %session.conn_id(), ?txn_reads, "using txn read holds for materialized view");
+
+            // We assume that earlier holds appear first in the txn holds.
```

Comment:
This sounds scary to me, I don't see anywhere around `self.txn_read_holds` that guarantees/enforces this. I also don't see anywhere around `self.txn_read_holds` that guarantees that the set of IDs in each `ReadHolds` is the same. It seems possible that we end up with a disjoint set of IDs in each `ReadHolds`. Though I can't at all remember how `self.txn_read_holds` is actually used in the wild.

Reply:
It's probably best to have @ggevay look at this file since he's the most familiar with materialized views and the read holds involved.

Reply:
Yeah, I had a feeling you'd say that when I wrote this code. I'll change it to:

1. Check that all read holds are for exactly the same IDs
2. Pick a least valid read across all the read holds

Plus, I'll try and remind myself why it is that we sometimes don't have read holds at this juncture. I once knew this, but it has escaped me since. :see_no_evil:

Reply:
Wait, if you don't do this entire if, but just simply call `acquire_read_holds`, it would take into account those earlier read holds, no? See my above comment.

Reply:
> Plus, I'll try and remind myself why it is that we sometimes don't have read holds at this juncture. I once knew this, but it has escaped me since. 🙈

You mean in `txn_read_holds`? It's because we put read holds in `txn_read_holds` only for refresh MVs.

Reply:
Looks like we were typing at the same time: https://github.com/MaterializeInc/materialize/pull/26620#discussion_r1562692230 :sweat_smile:

Reply:
> They would in the current implementation, but not in what we end up after my refactoring of the storage controller.

So, you mean that in that future refactoring, the implied capabilities would no longer include read holds?

Reply:
Ok, I see, thanks for the explanation!

Reply:
my pleasure! :bow: